src/combinatorics/simple_graph.lean

```diff
@@ -0,0 +1,217 @@
+
+/--
+A simple graph is an irreflexive symmetric relation `adj` on a vertex type `V`.
+The relation describes which pairs of vertices are adjacent.
+There is exactly one edge for every pair of adjacent edges;
+see `simple_graph.E` for the corresponding type of edges.
+-/
+structure simple_graph (V : Type u) :=
+(adj : V → V → Prop)
+(sym : symmetric adj . obviously)
+(loopless : irreflexive adj . obviously)
+
+/--
+The complete graph on a type `V` is the simple graph with all pairs of distinct vertices adjacent.
+-/
+def complete_graph (V : Type u) : simple_graph V :=
+{ adj := ne }
+
+instance (V : Type u) : inhabited (simple_graph V) :=
+⟨complete_graph V⟩
+
+instance complete_graph_adj_decidable (V : Type u) [decidable_eq V] :
+  decidable_rel (complete_graph V).adj :=
+by { dsimp [complete_graph], apply_instance }
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V)
+
+/-- `G.neighbor_set v` is the set of vertices adjacent to `v` in `G`. -/
+def neighbor_set (v : V) : set V := set_of (G.adj v)
+
+lemma ne_of_edge {a b : V} (hab : G.adj a b) : a ≠ b :=
+by { rintro rfl, exact G.loopless a hab }
+
+/--
+The edges of G consist of the unordered pairs of vertices related by
+`G.adj`.  It is given as a subtype of the symmetric square.
+-/
+def E : Type u := {x : sym2 V // x ∈ sym2.from_rel G.sym}
+
+/-- Allows us to refer to a vertex being a member of an edge. -/
+instance has_mem : has_mem V G.E := { mem := λ v e, v ∈ e.val }
+
+/-- Construct an edge from its endpoints. -/
+def edge_of_adj {v w : V} (h : G.adj v w) : G.E := ⟨⟦(v,w)⟧, h⟩
+
+lemma mem_of_adj {v w : V} (h : G.adj v w) :
+  v ∈ G.edge_of_adj h := sym2.mk_has_mem v w
+
+lemma mem_of_adj_right {v w : V} (h : G.adj v w) :
+  w ∈ G.edge_of_adj h := sym2.mk_has_mem_right v w
+
+lemma adj_iff_exists_edge {v w : V} (hne : v ≠ w) :
+  G.adj v w ↔ ∃ (e : G.E), v ∈ e ∧ w ∈ e :=
+begin
+  split, { intro, use ⟦(v,w)⟧, assumption, refine ⟨(G.mem_of_adj _), (G.mem_of_adj_right _)⟩ },
+  rintro ⟨e, ⟨w', hve⟩, ⟨v', hew⟩⟩,
+  have : e.val = ⟦(v,w)⟧, { rw [hve, sym2.eq_iff] at hew ⊢, cc },
+  have key := e.property, rwa this at key,
+end
+
+variables {G}
+
+/-- Given an edge and one vertex incident on it, construct the other one. -/
+noncomputable def E.other (e : G.E) {v : V} (h : v ∈ e) : V :=
+by { have : v ∈ e.val, apply h, exact this.other }
+
+lemma E.other_mem (e : G.E) {v : V} (h : v ∈ e) : e.other h ∈ e :=
+begin
+  change sym2.mem.other h ∈ e.val, have := sym2.mem_other_spec h,
+  convert sym2.mk_has_mem_right _ _; tauto,
+end
+
+lemma E.other_ne (e : G.E) {v : V} (h : v ∈ e) : e.other h ≠ v :=
+begin
+  have key := e.property,
+  erw [← sym2.mem_other_spec h, sym2.eq_swap] at key,
+  exact G.ne_of_edge key,
+end
+
+attribute [irreducible] E.other
+variables (G)
+
+instance E.inhabited [inhabited {p : V × V | G.adj p.1 p.2}] : inhabited G.E :=
+⟨begin
+  rcases inhabited.default {p : V × V | G.adj p.1 p.2} with ⟨⟨x, y⟩, h⟩,
+  use ⟦(x, y)⟧, rwa sym2.from_rel_prop,
+end⟩
```

Comment:
What purpose does this serve?

Reply:
Without it, the linter says
```
/- TYPES ARE MISSING INHABITED INSTANCES: -/
#print simple_graph.E /- inhabited instance missing -/
```

Reply:
I see. This is a bit of a pain. @fpvandoorn, do you have advice here on how to balance satisfying the `inhabited` linter, without cluttering up the library with a meaningless result? Would it be better to for example show that the edge set for the complete graph on two vertices was inhabited?

I used to hate the inhabited linter, but I realised that often enough it was prompting me to write "the first non-trivial example" and that this was often a useful exercise (typically because it showed you how terrible the API you'd just written was...).

Reply:
We just changed `simple_graph.E` to be a set `simple_graph.edge_set`.  The `E` type will reappear later when we create the multigraph interface.  But for now, means there's no need for the inhabited instance anymore, so it has been removed.